// @magento/peregrine/dist/Router/Router.js
import { createElement, Component } from 'react';
import { BrowserRouter, Route } from 'react-router-dom';
import { string, func, object } from 'prop-types';
import MagentoRouteHandler from './MagentoRouteHandler';

export default class MagentoRouter extends Component {

    render() {
        const {
            using: Router,
            routerProps,
            apiBase,
            __tmp_webpack_public_path__
        } = this.props;

        return createElement(
            Router,
            routerProps,
            createElement(Route, {
                render: ({ location }) => createElement(MagentoRouteHandler, {
                    location: location,
                    apiBase: apiBase,
                    __tmp_webpack_public_path__: __tmp_webpack_public_path__
                })
            })
        );
    }
}
MagentoRouter.propTypes = {
    /* Can be BrowserRouter, MemoryRouter, HashRouter, etc */
    using: func,
    routerProps: object,
    apiBase: string.isRequired,
    __tmp_webpack_public_path__: string.isRequired
};
MagentoRouter.defaultProps = {
    using: BrowserRouter,
    routerProps: {}
};
