// @magento/pwa-buildpack/dist/ExtensionComponentWrap/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; } /**
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * Note: This component should be moved to `peregrine` repository.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * It lives in this repo currently because this package consumes it, and we
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * don't have private npm packages yet
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                */

var errorStyles = {
    background: '#ff1b1b',
    color: '#fff',
    fontWeight: 'bold',
    fontSize: '1em',
    fontFamily: 'sans-serif'
};

var MagentoExtensionBoundary = function (_React$Component) {
    _inherits(MagentoExtensionBoundary, _React$Component);

    function MagentoExtensionBoundary() {
        _classCallCheck(this, MagentoExtensionBoundary);

        var _this = _possibleConstructorReturn(this, (MagentoExtensionBoundary.__proto__ || Object.getPrototypeOf(MagentoExtensionBoundary)).call(this));

        _this.state = {
            hasError: false
        };
        return _this;
    }

    _createClass(MagentoExtensionBoundary, [{
        key: 'componentDidCatch',
        value: function componentDidCatch() {
            var replacedID = this.props.replacedID;

            console.error('An error occurred within a part of the React component tree ' + 'created by a Magento Extension. Look for the component registered ' + ('to replace the mid "%c' + replacedID + '%c" to debug further.'));
            this.setState({ hasError: true });
        }
    }, {
        key: 'render',
        value: function render() {
            if (!this.state.hasError) {
                return this.props.children;
            }

            return _react2.default.createElement(
                'div',
                { style: errorStyles },
                'An error occurred within a Magento Extension. See the browser\'s JavaScript console for further details.'
            );
        }
    }]);

    return MagentoExtensionBoundary;
}(_react2.default.Component);

exports.default = MagentoExtensionBoundary;
